### ISSUE TITLE
panic: runtime error: slice bounds out of range

### ISSUE THREAD

**Issue by freeseacher** [influxdata/influxdb#6958]
### Bug report

__System info:__ RHEL 7.2 influxdb 0.12.2-1

__Possible Steps to reproduce:__
* today during migration host vmware moved time from 16:08 to 19:12. that time was fixed tha same way at about 16:34.

at 21:00 on regular migration influxdb got panic
```
[tsm1] 2016/07/04 21:00:13 compacting level 2 group (0) /var/lib/influxdb/data/noc/d90/153/000000230-000000002.tsm (#0)
[tsm1] 2016/07/04 21:00:13 compacting level 2 group (0) /var/lib/influxdb/data/noc/d90/153/000000232-000000002.tsm (#1)
panic: runtime error: slice bounds out of range

goroutine 51945911 [running]:
github.com/influxdata/influxdb/tsdb/engine/tsm1.(*indirectIndex).UnmarshalBinary(0xc617109d00, 0x7f91558e4319, 0xa73e4f6, 0xa73e4fe, 0x0, 0x0)
        /root/go/src/github.com/influxdata/influxdb/tsdb/engine/tsm1/reader.go:684 +0x466
github.com/influxdata/influxdb/tsdb/engine/tsm1.(*mmapAccessor).init(0xc2ceec5b30, 0x0, 0x0, 0x0, 0x0)
        /root/go/src/github.com/influxdata/influxdb/tsdb/engine/tsm1/reader.go:981 +0x4e8
github.com/influxdata/influxdb/tsdb/engine/tsm1.NewTSMReaderWithOptions(0x0, 0x0, 0xc417984368, 0x0, 0x0, 0x0)
        /root/go/src/github.com/influxdata/influxdb/tsdb/engine/tsm1/reader.go:158 +0x2e5
github.com/influxdata/influxdb/tsdb/engine/tsm1.(*Compactor).compact(0xc28c89ba40, 0xc29e09cc01, 0xc48d021460, 0x2, 0x2, 0x0, 0x0, 0x0, 0x0, 0x0)
        /root/go/src/github.com/influxdata/influxdb/tsdb/engine/tsm1/compact.go:459 +0x343
github.com/influxdata/influxdb/tsdb/engine/tsm1.(*Compactor).CompactFast(0xc28c89ba40, 0xc48d021460, 0x2, 0x2, 0x0, 0x0, 0x0, 0x0, 0x0)
        /root/go/src/github.com/influxdata/influxdb/tsdb/engine/tsm1/compact.go:486 +0x84
github.com/influxdata/influxdb/tsdb/engine/tsm1.func·007(0x0, 0xc48d021460, 0x2, 0x2)
        /root/go/src/github.com/influxdata/influxdb/tsdb/engine/tsm1/engine.go:570 +0x59f
created by github.com/influxdata/influxdb/tsdb/engine/tsm1.(*Engine).compactTSMLevel
        /root/go/src/github.com/influxdata/influxdb/tsdb/engine/tsm1/engine.go:596 +0x34a
```
and cyclic reboot after it. 
after several reboots by hand i found tsm file that caused that behavior
```
influx_inspect dumptsmdev 000000230-000000002.tsm
panic: runtime error: slice bounds out of range

goroutine 1 [running]:
github.com/influxdata/influxdb/tsdb/engine/tsm1.(*indirectIndex).UnmarshalBinary(0xc208062000, 0x7f84a10c0319, 0xa73e4f6, 0xa73e4fe, 0x0, 0x0)
	/root/go/src/github.com/influxdata/influxdb/tsdb/engine/tsm1/reader.go:684 +0x466
github.com/influxdata/influxdb/tsdb/engine/tsm1.(*mmapAccessor).init(0xc2080303c0, 0x0, 0x0, 0x0, 0x0)
	/root/go/src/github.com/influxdata/influxdb/tsdb/engine/tsm1/reader.go:981 +0x4e8
github.com/influxdata/influxdb/tsdb/engine/tsm1.NewTSMReaderWithOptions(0x0, 0x0, 0xc20802e028, 0x0, 0x0, 0x0)
	/root/go/src/github.com/influxdata/influxdb/tsdb/engine/tsm1/reader.go:158 +0x2e5
main.cmdDumpTsm1dev(0xc20800aea0)
	/root/go/src/github.com/influxdata/influxdb/cmd/influx_inspect/tsm.go:475 +0x1fc
main.main()
	/root/go/src/github.com/influxdata/influxdb/cmd/influx_inspect/main.go:115 +0xf82
```

1. is it known bug? 
2. does that file have any interest in debug purposes ?
3. can i restore data from it or even understand what metrics in what range was affected 
4. i am very disappointed with self healing possibility on influxdb and cyclic reboots. can we do something with it? 
5. it looks like all data that are behind 21:00 is unaccessible. it looks like there are no data. but of cause it exists, how can i access it ?

**Comment by freeseacher** [comment]
Any ideas? especially for how to get access for data.

**Comment by freeseacher** [comment]
Guys?

**Comment by gunnaraasen** [comment]
@freeseacher Can you check if the database will start when using the latest nightly? There have been a number of bug fixes to the TSM engine since 0.12.2 which may have addressed this panic.

InfluxDB should start up if you remove the problematic TSM file. Then you can run  `SHOW SHARDS` to identify the shard ID for the corrupted TSM file and the corresponding time range covered by the shard. Only that data in that database, retention policy, and time range should be inaccessible.

**Comment by freeseacher** [comment]
ok. i am done so. currently i am on latest nightly. 
`
Jul 08 15:33:32 db01 influxd[6259]: [run] 2016/07/08 15:33:32 InfluxDB starting, version 1.1.0~n201607080800, branch master, commit 11d141ae1fadfcac2a36e23d4aa38b36271a5daf
`
my big analytics db on latest nightly. God help me.

**Comment by freeseacher** [comment]
is there a way to dump data from currently broken files and restore them to db ?

**Comment by freeseacher** [comment]
@gunnaraasen 
https://github.com/influxdata/influxdb/issues/6985
https://github.com/influxdata/influxdb/issues/6984
https://github.com/influxdata/influxdb/issues/6982
https://github.com/influxdata/influxdb/issues/6981
https://github.com/influxdata/influxdb/issues/6979

no. latest nightly is not ready for testing of any kind at all.

**Comment by jwilder** [comment]
@freeseacher Can you elaborate more on your setup?  Specifically, you mentioned something about "regular vmware migration".   What kind of disks, hardware, kernel, etc.. are you using?  

The issues you have logged all appear to be some kind of incompatibility with the go runtime or possible hardware incompatibilities AFAICT.

* #6985 - `fatal error: unexpected signal during runtime execution` - in the go runtime `string` funcs where we're parsing a series keys from a MMAP'd tsm file.
* #6984 - `panic: runtime error: slice bounds out of range` - is triggered when we are decoding a compressed float block in an MMAP'd tsm file during a compactions.  Somehow the blocks has been truncated in the middle of compactions which makes no sense because we never modify tsm files after they are successfully written.
* #6982 - `panic: count of short block: got 0, exp 1` - is triggered when trying to examine a block on disk from a MMAP'd tsm file.  Somehow the block is missing a block type indicator.
* #6981 - `panic: runtime error: invalid memory address or nil pointer dereference` - is triggered in the middle of a compaction when writing the index portion of a TSM file.
* #6979 - looks to be the same issue as #6984.

We use memory mapped files for TSM.  It's possible our use of memory mapped files, your environment or the `vmware migration` are not compatible with each other.  Another thought is the the migration is somehow corrupting TSM files.

**Comment by freeseacher** [comment]
currently i already dropped all debug, config and the db.
but it seems that first trouble was related to migration from one hypervizor to another one. 
situation with it appears to be some kind of that
we are using such scheme
* special collector pulls metrics from network equipment and pushes it to nsq with line protocol.
* pmwriter, another our soft takes data from nsq and pushes it to influxdb, in a comfort for influxdb rate

collectors runs on another machine than pmwriter. influxdb runs on bare metal

So machine with collector was migrated from one hypervizor to another. and vmware tools on host changes clock time according to new hypervizor. 
but at the new hypervizor time was not in sync. so we got skip for 3 hour to the future. it last for about 30 minutes. than time at the new hypervizor was tunned to be in sync with real one, and vmware tools sync it to the host.
We got rate something about 5k metrics per second. 
it seems they was written to influx with wrong time. 
so after several hours we got another metrics with the same timestapms. 
and several hours later we got compact procedure. 

after that we got first error. 

than i had several days of fight with influxdb to achieve any stability. and at the end server reboot with memory error in mind. 
it seems that helped.
